proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+3) The notion that functional programming is "bad at cryptography" (ie, underperformant and has insurmountable challenges) is a self-fulfilling prophecy, and will remain so unless we invest in enough effort to disprove it.
+
+4) A fair lot of cryptographic projects in recent years has been specifically towards *cryptocurrencies*, rather than *cryptography in general*, and their results are at large fairly useless if they cannot be used outside of their ecosystems.
+
+All of these reasons have contributed to the stagnant state of Haskell cryptography, and have make it an unfavorable environment for developers of cryptosystems and distributed systems. However, none of these are actual blockers, and each of these things can be corrected with effort. There is much that can be done to improve the Haskell cryptography ecosystem, ranging from better low-level support, to higher-level abstractions:
+
+- Up-to-date algorithms
+- Hierarchy of typeclasses
+- Monadic interfaces
+- Safety abstractions
+- Type-level composition
+- Advanced data structure (eg, merkle trees)
+- Use of Generic for crypto-serialization
+
+As a result, there is much to be gained from investing in foundational Haskell cryptography. Haskell is well-suited to providing many aspects of information security through use of monadic interfaces and the type system, if the lower-level challenges of strictness and performant primitives can be solved. The community needs a cryptography library that is easily maintained, with minimal dependencies / baggage, and with documentation, tutorials, and unit tests to ease development.
+
+A first step would be developing bindings to a suitably-licensed open-source C or C++ library containing modern and up-to-date cryptography algorithms, including post-quantum cryptography schemes.
+
+There are [several C/C++ library candidates](https://en.wikipedia.org/wiki/Comparison_of_cryptography_libraries):
+
+- Botan
+- libgcrypt
+- Crypto++
+- libsodium / NaCl
+- NSS
+
+Of these, Botan was selected for several reasons:
+
+- It has a BSD license
+- It has a C FFI
+- It offers a broad variety of functionality and algorithms, including post-quantum cryptography
+- It has been [audited](https://botan.randombit.net/releases/audit_1.11.18.pdf) in the past
+- There are dead bindings targeting Botan 2 / GHC 8 that can be examined / revived
+- Significant implementation work has already been performed in order to validate the feasibility of Botan
+
+The others were not selected for several reasons:
+
+- libgcrypt is C so it requires no FFI, but it has a GNU LGPL license, and provides less functionality than Botan
+- Crypto++ has a Boost license (similar to BSD), but is C++ and has no C FFI, and provides less functionality than Botan
+- libsodium / NaCL does not provide the necessary functionality and there are several Haskell libsodium bindings already
```

Comment:
What's missing from `libsodium` for realistic industry use cases? The existence of existing bindings seems to be a reason to just go use them, rather than make new ones - a few examples of relevant primitives missing from `libsodium` would be handy here.

Reply:
This has been covered in [another comment](https://github.com/haskellfoundation/tech-proposals/pull/57#issuecomment-1734396606)